phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/CompactionScanner.java

```diff
@@ -2213,56 +2213,229 @@ class PhoenixLevelRowCompactor {
+                                    retainedCells.add(cell);
+                                }
+                            }
+                            cell = result.get(index + 1);
+                            if (!CellUtil.matchingColumn(cell, currentColumnCell)) {
+                                continue top;
+                            }
+                            index = index + 1;
+                        }
+                    }
+                }
+                if (deleteFamilyCell != null && deleteFamilyCell.getTimestamp() >= cell.getTimestamp()) {
+                    // This column is deleted by a delete family marker. Skip this column
+                    if (cell.getType() != Cell.Type.Put) {
+                        if (cell.getTimestamp() == maxLookbackWindowStart) {
+                            lastRowVersion.add(cell);
+                        } else if (!major) {
+                            // Retain the delete markers if the compaction is not major
+                            retainedCells.add(cell);
+                        }
+                    }
+                    index = skipColumn(result, currentColumnCell, retainedCells, index);
+                    continue top;
+                }
+                // Process back-to-back deleted cell versions. Phoenix currently does not use delete cell version
+                // markers. This processing should not happen and is added for completeness
+                while (cell.getType() == Cell.Type.Delete) {
+                    if (cell.getTimestamp() == maxLookbackWindowStart) {
                         lastRowVersion.add(cell);
+                    } else if (!major) {
+                        retainedCells.add(cell);
+                    }
+                    Cell nextCell = result.get(index + 1);
+                    if (!CellUtil.matchingColumn(currentColumnCell, nextCell)) {
+                        continue top;
+                    }
+                    if (nextCell.getType() == Cell.Type.Put && cell.getTimestamp() == nextCell.getTimestamp()) {
+                        // This put cell is masked by the delete marker
+                        index = index + 1;
+                        cell = nextCell;
+                    }
+                }
+                if (cell.getType() == Cell.Type.DeleteColumn) {
+                    // The rest of the column is masked by this delete column cell
+                    if (cell.getTimestamp() == maxLookbackWindowStart) {
+                        lastRowVersion.add(cell);
+                    } else if (!major) {
+                        retainedCells.add(cell);
+                    }
+                    index = skipColumn(result, currentColumnCell, retainedCells, index);
+                    continue top;
+                }
+                if (cell.getType() == Cell.Type.Put) {
+                    lastRowVersion.add(cell);
+                    if (ScanUtil.isEmptyColumn(cell, emptyCF, emptyCQ)) {
```

Comment:
We do gap analysis for TTL processing and we expire rows only when the current compaction is a major compaction and so we should not track empty column cells when it is not a major compaction